fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerByRecords.java

```diff
@@ -0,0 +1,677 @@
+        } else {
+            // Both ends inclusive.
+            return new TupleRange(low, high, EndpointType.RANGE_INCLUSIVE, EndpointType.RANGE_INCLUSIVE);
+        }
+    }
+
+    /**
+     * Builds (transactionally) the endpoints of an index. What this means is that builds everything from the beginning of
+     * the key space to the first record and everything from the last record to the end of the key space.
+     * There won't be any records within these ranges (except for the last record of the record store), but
+     * it does mean that any records in the future that get added to these ranges will correctly update
+     * the index. This means, e.g., that if the workload primarily adds records to the record store
+     * after the current last record (because perhaps the primary key is based off of an atomic counter
+     * or the current time), running this method will be highly contentious, but once it completes,
+     * the rest of the index build should happen without any more conflicts.
+     *
+     * This will return a (possibly null) {@link TupleRange} that contains the primary keys of the
+     * first and last records within the record store. This can then be used to either build the
+     * range right away or to then divy-up the remaining ranges between multiple agents working
+     * in parallel if one desires.
+     *
+     * @param store the record store in which to rebuild the index
+     * @param recordsScanned continues counter
+     * @return a future that will contain the range of records in the interior of the record store
+     */
+
+    @Nonnull
+    public CompletableFuture<TupleRange> buildEndpoints(@Nonnull FDBRecordStore store,
+                                                        @Nullable AtomicLong recordsScanned) {
+        final RangeSet rangeSet = new RangeSet(store.indexRangeSubspace(common.getIndex()));
+        if (TupleRange.ALL.equals(recordsRange)) {
+            return buildEndpoints(store, rangeSet, recordsScanned);
+        }
+        // If records do not occupy whole range, first mark outside as built.
+        final Range asRange = recordsRange.toRange();
+        return CompletableFuture.allOf(
+                rangeSet.insertRange(store.ensureContextActive(), null, asRange.begin),
+                rangeSet.insertRange(store.ensureContextActive(), asRange.end, null))
+                .thenCompose(vignore -> buildEndpoints(store, rangeSet, recordsScanned));
+    }
+
+    @Nonnull
+    private CompletableFuture<TupleRange> buildEndpoints(@Nonnull FDBRecordStore store, @Nonnull RangeSet rangeSet,
+                                                         @Nullable AtomicLong recordsScanned) {
+        boolean isIdempotent = store.getIndexMaintainer(common.getIndex()).isIdempotent();
+        final IsolationLevel isolationLevel =
+                isIdempotent ?
+                // If idempotent: since double indexing is harmless, we can use individual records protection instead of
```

Comment:
This change gotten in a few months ago. I'm not sure what's the "split records" implications.

Reply:
@jjezra "Split records" means storing a record with more than one key-value pair (because of the record value is large)

@ScottDugas Good point. I didn't check how w/r split records was implemented so I don't have an answer yet. But even if it's not an issues, I think we can at least a test to prove it. Created https://github.com/FoundationDB/fdb-record-layer/issues/1093

Reply:
Is there any way to change one of the value's continuation records without changing the first one? I assume that all the records of a single value are written in one transaction.